File: Code Examples/Medium/RuntimeErrorExample.swift
class RuntimeErrorExample {
    static func main() {
        var data = [5, 2, 9, 1, 5, 6]
        quickSort(&data, left: 0, right: data.count) // <- ошибка: выход за границу
        print("Sorted:")
        for val in data {
            print(val)
        }
    }

    static func quickSort(_ arr: inout [Int], left: Int, right: Int) {
        if left >= right { return }

        let pivot = arr[(left + right) / 2]
        let index = partition(&arr, left: left, right: right, pivot: pivot)
        quickSort(&arr, left: left, right: index - 1)
        quickSort(&arr, left: index, right: right)
    }

    static func partition(_ arr: inout [Int], left: Int, right: Int, pivot: Int) -> Int {
        var left = left
        var right = right

        while left <= right {
            while arr[left] < pivot { left += 1 }
            while arr[right] > pivot { right -= 1 } // <- тут произойдёт выход за границу

            if left <= right {
                arr.swapAt(left, right)
                left += 1
                right -= 1
            }
        }
        return left
    }
}

File: Code Examples/Medium/RuntimeErrorExample.cs
//quick-sort example
class RuntimeErrorExample
{
    static void Main()
    {
        int[] data = { 5, 2, 9, 1, 5, 6 };
        QuickSort(data, 0, data.Length); // wrong bounds on purpose
        Console.WriteLine("Sorted:");
        foreach (int val in data)
            Console.WriteLine(val);
    }

    static void QuickSort(int[] arr, int left, int right)
    {
        if (left >= right)
            return;

        int pivot = arr[(left + right) / 2];
        int index = Partition(arr, left, right, pivot);
        QuickSort(arr, left, index - 1);
        QuickSort(arr, index, right);
    }

    static int Partition(int[] arr, int left, int right, int pivot)
    {
        while (left <= right)
        {
            while (arr[left] < pivot) left++;
            while (arr[right] > pivot) right--; // here we can get desired runtime error (outofboundsexception)

            if (left <= right)
            {
                int tmp = arr[left];
                arr[left] = arr[right];
                arr[right] = tmp;
                left++;
                right--;
            }
        }
        return left;
    }
}
